spec/datadog/core/utils/safe_dup_spec.rb

```diff
@@ -0,0 +1,55 @@
+require 'datadog/core/utils/safe_dup'
+
+RSpec.describe Datadog::Core::Utils::SafeDup do
+  describe '.frozen_or_dup' do
+    context 'when given a frozen string' do
+      it 'returns the original input' do
+        input = 'a_frozen_string'.freeze
+
+        result = described_class.frozen_or_dup(input)
+
+        expect(result).to eq(input)
+        expect(result).to equal(input)
+        expect(result).to be_frozen
+      end
+    end
+
+    context 'when given a string' do
+      it 'returns a duplicate' do
+        input = 'a_string'
+
+        result = described_class.frozen_or_dup(input)
+
+        expect(result).to eq(input)
+        expect(result).not_to equal(input)
+        expect(result).not_to be_frozen
+      end
+    end
+  end
+
+  describe '.frozen_dup' do
+    context 'when given a frozen string' do
+      it 'returns the original input' do
+        input = 'a_frozen_string'.freeze
+
+        result = described_class.frozen_dup(input)
+
+        expect(result).to eq(input)
+        expect(result).to equal(input)
+        expect(result).to be_frozen
+      end
+    end
+
+    context 'when given a string' do
+      it 'returns a frozen duplicate' do
+        input = 'a_string'
+
+        result = described_class.frozen_dup(input)
+
+        expect(result).to eq(input)
+        expect(result).not_to equal(input)
```

Comment:
Minor: Suggestion to match my other suggestion above

```suggestion
        expect(result).not_to be(input)
```